src/main/java/org/mockito/internal/invocation/MatcherApplicationStrategy.java

```diff
@@ -91,21 +93,35 @@ private static MatcherApplicationType getMatcherApplicationType(Invocation invoc
         final int expandedArguments = invocation.getArguments().length;
         final int matcherCount = matchers.size();
 
+        boolean treatVarargsAsArray = false;
```

Comment:
Wow. I'm very impressed that you were able to inject the new algorithm here. The code is pretty complex around here. I'd love to see it refactored at some point in the future.

It is complex but you have added fantastic test coverage that I was not able to break down :)